### PR TITLE
LibJS: Add Array.prototype.splice

### PR DESCRIPTION
**Pull request by Lubrsi**
Adds splice to Array.prototype according to the specification:
https://tc39.es/ecma262/#sec-array.prototype.splice

I'm not sure what to do about the casts on lines 663, 665 and 678. It isn't pretty :(

### CODE REVIEW

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 1]
```diff
@@ -642,4 +643,95 @@ Value ArrayPrototype::every(Interpreter& interpreter)
...
+    if (insert_count < actual_delete_count) {
+        for (size_t i = actual_start; i < initial_length - actual_delete_count; ++i) {
+            auto from = this_object->get_by_index(i + actual_delete_count);
+            if (interpreter.exception())
+                return {};
+
+            auto to = i + insert_count;
+
+            if (!from.is_empty())
+                this_object->put_by_index(to, from);
```
Comment (linusg) [inline Libraries/LibJS/Runtime/ArrayPrototype.cpp:705]: Missing exception check

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 2]
```diff
@@ -642,4 +643,95 @@ Value ArrayPrototype::every(Interpreter& interpreter)
...
+    } else if (insert_count > actual_delete_count) {
+        for (size_t i = initial_length - actual_delete_count; i > actual_start; --i) {
+            auto from = this_object->get_by_index(i + actual_delete_count - 1);
+            if (interpreter.exception())
+                return {};
+
+            auto to = i + insert_count - 1;
+
+            if (!from.is_empty())
+                this_object->put_by_index(to, from);
```
Comment (linusg) [inline Libraries/LibJS/Runtime/ArrayPrototype.cpp:721]: Missing exception check

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 3]
```diff
@@ -642,4 +643,95 @@ Value ArrayPrototype::every(Interpreter& interpreter)
...
+            else
+                this_object->delete_property(PropertyName(to));
+        }
+    }
+
+    for (size_t i = 0; i < insert_count; ++i) {
+        this_object->put_by_index(actual_start + i, interpreter.argument(i + 2));
```
Comment (linusg) [inline Libraries/LibJS/Runtime/ArrayPrototype.cpp:728]: Missing exception check

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 4]
```diff
@@ -642,4 +643,95 @@ Value ArrayPrototype::every(Interpreter& interpreter)
...
+    }
+
+    auto new_length_value = Value((i32)new_length);
+    this_object->put("length", new_length_value);
```
Comment (linusg) [inline Libraries/LibJS/Runtime/ArrayPrototype.cpp:732]: Missing exception check

Comment (linusg) [reply Libraries/LibJS/Runtime/ArrayPrototype.cpp:732]: Also I think that's overly verbose and could just be

```cpp
this_object->put("length", Value((i32)new_length));
```

**Comment by Lubrsi** [comment]
Should be resolved now 👍

**Pull request merged**